File: 2024/01/main.py
from collections import Counter


class Solution:
    def part1(self):
        with open("2024/01/input.txt") as file:
            left, right = zip(*(map(int, line.split()) for line in file))

        left_sorted = sorted(left)
        right_sorted = sorted(right)

        return sum(abs(l - r) for l, r in zip(left_sorted, right_sorted))

    def part2(self):
        with open("2024/01/input.txt") as file:
            left, right = zip(*(map(int, line.split()) for line in file))

        right_counter = Counter(right)

        return sum(l * right_counter[l] for l in left)


if __name__ == "__main__":
    s = Solution()
    print(s.part1())
    print(s.part2())

File: 2024/01/main.rb
require 'set'

class Solution
  def initialize(file_path)
    @file_path = file_path
  end

  def part1
    left, right = read_and_split
    left_sorted = left.sort
    right_sorted = right.sort

    left_sorted.zip(right_sorted).sum { |l, r| (l - r).abs }
  end

  def part2
    left, right = read_and_split
    right_count = right.tally

    left.sum { |l| l * right_count.fetch(l, 0) }
  end

  private

  def read_and_split
    left = []
    right = []
    File.foreach(@file_path) do |line|
      l, r = line.split.map(&:to_i)
      left << l
      right << r
    end
    [left, right]
  end
end

solution = Solution.new('2024/01/input.txt')
puts solution.part1
puts solution.part2
